[2025-07] Automate CalVer branch detection for quarterly releases

juanpprieto opened the pull request:
# Automated CalVer Release System with Major Version Protection

### WHY are these changes introduced?

Fixes the critical version-API mismatch issue where 2025-07 Storefront API changes would incorrectly release as version 2025.5.x instead of 2025.7.0, breaking the calendar versioning alignment with Shopify's API release schedule.

**Problem:** The current manual release process requires maintainers to:
1. Manually determine the correct CalVer branch (2025-05, 2025-07, etc.) based on changesets
2. Manually prevent major changes from contaminating patch releases
3. Manually enforce CalVer formatting for Hydrogen packages
4. Risk releasing API updates under wrong version numbers

This creates opportunities for human error and version misalignment that could confuse developers about which Storefront API version they're using.

### WHAT is this pull request doing?

This PR introduces a comprehensive **Automated CalVer Release System** with **Major Version Protection** that:

#### 🤖 **Automated CalVer Branch Detection**
- Automatically determines correct release branches (2025-05, 2025-07, etc.) based on changeset analysis
- Detects when major changes require advancing to next quarter (Q3 = 2025-07)
- Eliminates manual branch selection and reduces human error

#### 🛡️ **Major Version Protection System**
- **Blocks major changes** from merging when patch/minor releases are pending
- Prevents version contamination (e.g., 2025-07 API changes releasing as 2025.5.1)
- Provides maintainer bypass mechanism (`/bypass-major-safeguard` comment)
- Runs as GitHub Actions check with clear blocking messages

#### 📅 **CalVer Enforcement**
- Automatically transforms changeset versions to CalVer format (YYYY.Q.P)
- Ensures `@shopify/hydrogen`, `@shopify/hydrogen-react`, and `skeleton` use calendar versioning
- Preserves semver for non-CalVer packages (`@shopify/cli-hydrogen`, `@shopify/create-hydrogen`, etc.)
- Fixed patch version bug: `2025.5.0` → `2025.5.1` (not `2025.5.0.1`)

#### 🗂️ **Consolidated Script Architecture**
- All CalVer/release scripts moved to `.changeset/` directory
- Shared utilities in `calver-shared.js` reduce code duplication by ~100 lines
- Modular design with clear separation of concerns

### HOW to test your changes?

#### 🧪 **Comprehensive Testing with Tophat**

This PR includes a complete `tophat.md` document with 7 validation sections. Run these tests to verify functionality:

##### **1. Test CalVer Utilities and CLI Commands**
```bash
# Test version parsing and next version calculation
node .changeset/calver-shared.js get-next "2025.5.0" "major"   # Should output: 2025.7.0
node .changeset/calver-shared.js get-next "2025.5.0" "minor"   # Should output: 2025.5.1
node .changeset/calver-shared.js get-next "2025.5.0" "patch"   # Should output: 2025.5.1

# Test bump type detection
node .changeset/calver-shared.js detect-bump "2025.5.0" "2025.7.0"  # Should output: major
node .changeset/calver-shared.js detect-bump "2025.5.0" "2025.5.1"  # Should output: minor

# Test CalVer package listing
node .changeset/calver-shared.js list-calver-packages  # Should list: @shopify/hydrogen @shopify/hydrogen-react skeleton
```

##### **2. Test Branch Detection Logic**
```bash
# Test branch detection without major changesets (should use existing 2025-05)
node .changeset/get-calver-version-branch.js

# Create a test major changeset and test again
cat > .changeset/test-major-validation.md << 'EOF'
---
"@shopify/hydrogen": major
---
Test major changeset for validation
EOF

# Test with major changeset (should detect 2025-07 is needed)
node .changeset/get-calver-version-branch.js

# Clean up test changeset
rm .changeset/test-major-validation.md
```

##### **3. Test CalVer Enforcement (Dry Run)**
```bash
# Test dry run to see what would happen
npm run test:calver:dry

# Alternative: Direct script execution
node .changeset/enforce-calver-local.js --dry-run
```

##### **4. Test CalVer vs Standard Changeset Comparison**
```bash
# Full comparison test (non-destructive, reverts all changes)
npm run test:calver

# Alternative: Direct execution
./.changeset/test-calver-comparison.sh
```

Expected output should show:
- CalVer packages use YYYY.Q.P format (2025.5.0 → 2025.5.1)
- Non-CalVer packages use standard semver
- No files are permanently modified

##### **5. Test Major Version Protection System**
```bash
# Test 1: No major changes (should pass)
node .changeset/check-major-protection.js --local
echo "Expected: ✅ No Major Changes"

# Test 2: Create major changeset and test (should block)
cat > .changeset/test-major-block.md << 'EOF'
---
"@shopify/hydrogen": major
---
Test major that should be blocked
EOF

node .changeset/check-major-protection.js --local
cat .github/protection-result.json  # Should show "blocked": true

# Test 3: Create patch alongside major (should still block)
cat > .changeset/test-patch.md << 'EOF'
---
"@shopify/hydrogen": patch
---
Test patch for protection validation
EOF

node .changeset/check-major-protection.js --local
echo "Expected: 🚫 Major Changes Blocked"

# Clean up test changesets
rm .changeset/test-patch.md .changeset/test-major-block.md
rm -f .github/protection-result.json
```

##### **6. Test Workflow Configuration**
```bash
# Check workflow syntax (requires GitHub CLI)
if command -v gh &> /dev/null; then
    gh workflow list --repo Shopify/hydrogen | grep -E "(Major Protection|Test CalVer|Changesets)"
    echo "✅ Workflows are configured"
else
    echo "ℹ️  Install GitHub CLI to test workflows: gh auth login"
fi
```

##### **7. Test Complete Protection Workflow**
```bash
# Scenario 1: Test with mixed changesets (major + patch)
cat > .changeset/test-major-mixed.md << 'EOF'
---
"@shopify/hydrogen": major
"@shopify/mini-oxygen": patch
---
Test mixed major and patch changes
EOF

node .changeset/check-major-protection.js --local
echo "Expected: Should block because of major change to protected package"

# Scenario 2: Test with major change to non-protected package (should pass)
cat > .changeset/test-major-allowed.md << 'EOF'
---
"@shopify/create-hydrogen": major
"@shopify/mini-oxygen": patch
---
Test major change to non-protected package
EOF

rm .changeset/test-major-mixed.md  # Remove the blocking changeset
node .changeset/check-major-protection.js --local
echo "Expected: Should pass - major change to non-protected package is allowed"

# Scenario 3: Test CalVer version calculation with major changesets
echo "Testing CalVer behavior with major changesets:"
node .changeset/calver-shared.js get-next "2025.5.0" "major"  # Should be 2025.7.0
echo "Major changesets advance to next quarter (2025.5.0 → 2025.7.0)"

# Clean up
rm .changeset/test-major-allowed.md
rm -f .github/protection-result.json
```

#### 🔄 **Integration Testing**
Create a test branch to validate CI workflows:
```bash
git checkout -b test-calver-validation
echo "# Test file" > test-validation.md
git add test-validation.md
git commit -m "test: validate CalVer protection system"
git push -u origin test-calver-validation

# Create PR and observe:
# 1. Major Protection workflow should run
# 2. Test CalVer workflow should run (if CalVer files changed)
# 3. Protection should pass (no major changesets in test)
```

#### Post-merge steps

**Documentation Updates Required:**
- [ ] Update `docs/changelog.json` after releases to enable `h2 upgrade` command
- [ ] Update `.github/workflows/changesets.yml` line 32 when moving to new major version (quarterly)
- [ ] Monitor first few releases to ensure CalVer enforcement works correctly

#### Checklist

- [x] I've read the [Contributing Guidelines](https://github.com/Shopify/hydrogen/blob/main/CONTRIBUTING.md)
- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've added a [changeset](https://github.com/Shopify/hydrogen/blob/main/CONTRIBUTING.md#changesets) if this PR contains user-facing or noteworthy changes
- [x] I've added [tests](https://github.com/Shopify/hydrogen/blob/main/CONTRIBUTING.md#testing) to cover my changes
- [x] I've added or updated the documentation

### 🔧 **Technical Implementation Details**

#### Files Added/Modified:
```
.changeset/
├── calver-shared.js                     # Core CalVer utilities (275 lines)
├── enforce-calver-ci.js                 # CI CalVer enforcement (refactored)
├── enforce-calver-local.js              # Local CalVer testing (refactored)
├── get-calver-version-branch.js         # Automatic branch detection
├── check-major-protection.js            # Major version protection logic
├── changeset-protection-utils.js        # Shared protection utilities
├── process-major-version-bypass.js      # Maintainer bypass handler
└── test-calver-comparison.sh            # Comprehensive testing script

.github/workflows/
├── major-protection.yml                 # New: Major protection workflow
└── test-calver.yml                      # New: CalVer testing workflow

docs/
└── CALVER.md                           # Complete system documentation

tophat.md                               # Validation guide for maintainers
```

juanpprieto committed:
Automate CalVer branch detection for quarterly releases

Eliminates manual quarterly updates to latestBranch in changesets.yml by
implementing automatic branch detection based on package versions and changesets.

Key changes:
- Add get-calver-version-branch.js to detect correct release branch
- Add calver-shared.js for reusable CalVer logic
- Update changesets.yml to use automated detection
- Add safeguards to prevent PR conflicts when release PR is open
- Fix enforce-calver-local.js dry-run to read actual changesets

Fixes #3110

juanpprieto committed:
feat: implement Major Version Protection system

Adds protection to prevent major changesets from contaminating patch/minor releases.
This ensures users can get critical bug fixes without forced major upgrades.

Key components:
- major-protection.yml: Combined workflow for protection checks and bypass
- changeset-protection-utils.js: Shared utilities (86 lines)
- check-major-protection.js: Protection logic (79 lines)
- process-major-version-bypass.js: Bypass command handler (43 lines)
- Updated CALVER.md with Major Version Protection documentation

Total implementation: 270 lines (71% reduction from original approach)

Protects packages: @shopify/hydrogen, @shopify/hydrogen-react, skeleton
Bypass mechanism: Maintainers can comment /bypass-major-safeguard

juanpprieto committed:
refactor: use calver-shared utilities to reduce code duplication

Both enforce-calver-ci.js and enforce-calver-local.js now import shared
functions from calver-shared.js, eliminating ~100 lines of duplicate code

juanpprieto committed:
refactor: consolidate all CalVer/release scripts to .changeset directory

- Move protection scripts from .github/scripts/ to .changeset/
- Move test-calver-comparison.sh from scripts/ to .changeset/
- Update all workflow and package.json references
- Fix CalVer patch logic (2025.5.0 → 2025.5.1, not 2025.5.0.1)
- Reduce code duplication by ~100 lines through shared utilities

juanpprieto committed:
feat: add --local flag to major protection script for maintainer testing

- Enable local testing of major version protection system
- Bypasses GitHub API calls when --local flag is used
- Allows maintainers to validate protection logic locally
- Essential for tophat validation workflows

kdaviduik commented on .changeset/changeset-protection-utils.js:
```diff
@@ -0,0 +1,87 @@
+// Check if PR has bypass label
+function hasBypassLabel(prNumber) {
+  const labels = gh('pr', 'view', prNumber, '--json', 'labels', '-q', '.labels[].name');
+  return labels && labels.includes('major-bypass-active');
```
I thought you didn't want any bypasses to exist at all?

juanpprieto replied:
Yeah this will be useful if / when we drift majors in the future. It's better to have a bit of control here

kdaviduik commented on .changeset/get-calver-version-branch.js:
```diff
@@ -0,0 +1,102 @@
+    // Check if we need to advance to next quarter
+    if (hasMajorChangesets()) {
+      // SAFEGUARD 2: Only advance if no open PR
+      console.error('Major changesets detected, advancing to next quarter');
```
Why are we using `console.error` for all of these instead of `console.log`?

juanpprieto replied:
copilot :(

juanpprieto committed:
fix: use proper console logging semantics in branch detection script

- console.log() for informational messages (PR found, major changesets detected)
- console.warn() for warnings (GitHub CLI issues)
- console.error() reserved for actual errors
